### DISCUSSION TITLE
On-demand Revalidation interval

### DISCUSSION THREAD

**Discussion by paragkatoch** [vercel/next.js#38513] [Ideas]
### Describe the feature you'd like to request

We can revalidate 1 path with res.revalidate(), which triggers a serverless function

```
// pages/api/revalidate.js

export default async function handler(req, res) {
    await res.revalidate('/path-to-revalidate')
    return res.json({ revalidated: true })
}
```
But, if a server is receiving 1 revalidation request every second, then this wouldn't be a cost effective approach.

### Describe the solution you'd like

This can be improved if Next.js can group all the On-Demand revalidation request and execute all of them after a certain interval of time. (batch and execute)

A new prop inside getStaticProps function to determine this interval 

```
export async function getStaticProps() {
  const res = await fetch('https://.../posts')
  const posts = await res.json()

  return {
    props: {
      posts,
    },

    // Next.js will attempt to process the page:
    // - When an on-demand revalidation request comes in
    // - At most once every 10 seconds
    on-demand-revalidate: 10, // In seconds
  }
}
```

### Describe alternatives you've considered

Alternative would be that the user implement this on their own i.e., grouping and executing res.revalidate() on an array of paths to revalidate in a time interval, but if that triggers multiple serverless functions then it would not make any difference

**Comment by balazsorban44** [comment]
Check out https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#using-on-demand-revalidation

Specifically this code:

```js
// Check for secret to confirm this is a valid request
if (req.query.secret !== process.env.MY_SECRET_TOKEN) {
  return res.status(401).json({ message: 'Invalid token' })
}
```

The recommended solution is to attach a secret to the revalidation request from your CMS and don't call `res.revalidate`, unless the secret matches. This way you can avoid anyone spamming your revalidation endpoint.

**Comment by paragkatoch** [comment]
@balazsorban44 You are totally correct. 

But what I want to highlight is that the current implementation of On Demand Revalidation is costly because NextJs will initiate a new serverless function for each revalidation request. 

If an application receives 10 revalidation requests in 1 second, NextJs will initiate 10 separate serverless functions; similarly, if an application receives 50 revalidation requests in 60 seconds, NextJs will initiate 50 serverless functions. 

If NextJs can group all revalidation requests and execute them with a single serverless function, then cost can be significantly reduced. We can add a new property called 'on-demand-revalidation-interval' that determines how long to wait before executing all revalidation requests.
